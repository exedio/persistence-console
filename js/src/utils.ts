export function delay<T>(
  call: Promise<T>,
  fail: boolean = false,
  timeout: number = 1000,
): Promise<T> {
  return new Promise((resolve, reject) => {
    setTimeout(() => {
      if (fail) {
        reject(new Error("utils.ts delay failed"));
      } else {
        call.then((r) => resolve(r));
      }
    }, timeout);
  });
}
