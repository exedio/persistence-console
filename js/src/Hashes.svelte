<script lang="ts">
  import { fly } from "svelte/transition";
  import "@/table-grey.css";
  import { get } from "@/api/api";
  import { type Hash as ApiHash } from "@/api/types";
  import { PromiseTracker } from "@/api/PromiseTracker.svelte.js";
  import PromiseTrackerReload from "@/api/PromiseTrackerReload.svelte";
  import { Hash } from "@/UseHashes.svelte";
  import { useWithStore } from "@/utils";

  const hashesStore = new Map<string, Hash>();

  const hashes = new PromiseTracker(() =>
    get<ApiHash[]>("hashes").then((response) =>
      useWithStore(
        hashesStore,
        toId,
        (source) => new Hash(source),
        (target, source) => target.update(source),
        response,
      ),
    ),
  );
  const errors: Error[] = $state([]);

  function measureAll(hashes: Hash[]) {
    let p = Promise.resolve();
    hashes.forEach((h) => (p = p.then(async () => await h.measure(errors))));
  }

  function toId(response: ApiHash): string {
    return response.type + "." + response.name;
  }
</script>

<table class="grey">
  <caption>
    Hashes
    <PromiseTrackerReload tracker={hashes} />
  </caption>
  <thead>
    <tr class="relative">
      <th rowspan="2">Type</th>
      <th rowspan="2">Name</th>
      <th colspan="2">Plain Text</th>
      <th colspan="2">Algorithm</th>
      <th rowspan="2" class="time">
        Time<small>/ns</small>
        {#await hashes.promise() then hashes}
          {#if hashes.length > 0}
            <button class="measure" onclick={() => measureAll(hashes)}
              >&#128336;
            </button>
          {/if}
        {:catch _error}
          &nbsp;
        {/await}
      </th>
    </tr>
    <tr>
      <th>Limit</th>
      <th>Validator</th>
      <th>ID</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
    {#await hashes.promise()}
      <tr>
        <td colspan="7" class="empty">fetching data</td>
      </tr>
    {:then hashes}
      {#each hashes as hash (toId(hash))}
        {@const measurement = hash.getMeasurement()}
        <tr class="relative">
          <td>
            <button class="hash" onclick={() => hash.toggle()}>
              {#if hash.isToggled()}
                &#8594;
              {:else}
                &#8595;
              {/if}
            </button>
            {hash.type}
          </td>
          <td>{hash.name}</td>
          <td class="number">{hash.plainTextLimit}</td>
          <td class:notAvailable={!hash.plainTextValidator}
            >{hash.plainTextValidator}</td
          >
          <td>{hash.algorithmID}</td>
          <td>{hash.algorithmDescription}</td>
          <td class="number">
            {#if measurement !== undefined}
              {measurement.toLocaleString("en-US")}
            {/if}
            <button class="measure" onclick={() => hash.measure(errors)}
              >&#128336;
            </button>
          </td>
        </tr>
        {#if hash.isToggled()}
          <tr in:fly={{ y: -10, duration: 200 }}>
            <td colspan="7" class="expansion">
              <input bind:value={hash.plainText} placeholder="Plain Text" />
              <button onclick={() => hash.computeHash(errors)}>Hash</button>
              <br />
              <small>{hash.getPlainTextHashed()}</small>
            </td>
          </tr>
        {/if}
      {:else}
        <tr>
          <td colspan="7" class="empty">There are no hashes in the model.</td>
        </tr>
      {/each}
    {:catch error}
      <tr>
        <td colspan="7" class="error">{error.message}</td>
      </tr>
    {/await}
  </tbody>
</table>

{#if errors.length > 0}
  <h2>Errors</h2>
  <ul>
    {#each errors as error}
      <li>{error}</li>
    {/each}
  </ul>
{/if}

<style>
  table {
    margin-left: 1em;
  }

  button.measure {
    position: absolute;
    top: 0;
    right: -25px;
    border: 0;
    background: white;
  }

  button.hash {
    position: absolute;
    top: 0;
    left: -20px;
    border: 0;
    background: white;
  }

  td.expansion {
    max-width: 40px;
  }

  td.expansion input {
    width: 90%;
  }

  tr.relative {
    position: relative;
  }

  th.time {
    min-width: 7em;
  }
</style>
