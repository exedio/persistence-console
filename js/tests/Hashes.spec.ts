import { mount } from "svelte";
import { default as Hashes } from "@/Hashes.svelte";
import {
  mockFetch,
  request,
  responseFailure,
  responseSuccess,
} from "@t/mockApi";
import { flushPromises, formatHtml } from "@t/utils";
import type {
  DoHashRequest,
  DoHashResponse,
  HashesResponse,
} from "@/api/types";

describe("Hashes", () => {
  it("should render an empty table", async () => {
    const mock = mockFetch();
    mock.mockResolvedValueOnce(responseSuccess([] satisfies HashesResponse[]));
    await mountComponent();
    expect(mock).toHaveBeenCalledExactlyOnceWith("/myApiPath/hashes");
    expect(await formatHtml(document.body)).toMatchSnapshot();
  });

  it("should render a non-empty table", async () => {
    const mock = mockFetch();
    mock.mockResolvedValueOnce(
      responseSuccess([
        {
          type: "myType",
          name: "myName",
          plainTextLimit: 150,
          plainTextValidator: "myValidator",
          algorithmID: "myAlgorithmID",
          algorithmDescription: "myAlgorithmDescription",
        },
        {
          type: "myTypeWithoutLimit",
          name: "myNameWithoutLimit",
          plainTextLimit: 155,
          plainTextValidator: undefined,
          algorithmID: "myAlgorithmIDWithoutLimit",
          algorithmDescription: "myAlgorithmDescriptionWithoutLimit",
        },
      ] satisfies HashesResponse[]),
    );
    await mountComponent();
    expect(mock).toHaveBeenCalledExactlyOnceWith("/myApiPath/hashes");
    expect(await formatHtml(document.body)).toMatchSnapshot();
  });

  it("should render a table body with all hashes measured", async () => {
    mockFetch().mockResolvedValueOnce(responseSuccess(threeHashes));
    await mountComponent();

    const mock = mockFetch();
    mock
      .mockResolvedValueOnce(
        responseSuccess({
          hash: "DO NOT ASSERT myHash1",
          elapsedNanos: 50010,
        } satisfies DoHashResponse),
      )
      .mockResolvedValueOnce(
        responseSuccess({
          hash: "DO NOT ASSERT myHash2",
          elapsedNanos: 50020,
        } satisfies DoHashResponse),
      )
      .mockResolvedValueOnce(
        responseSuccess({
          hash: "DO NOT ASSERT myHash3",
          elapsedNanos: 50030,
        } satisfies DoHashResponse),
      );
    (document.querySelectorAll(".measure").item(0) as HTMLElement).click();
    await flushPromises();
    expect(mock).toHaveBeenNthCalledWith(
      1,
      "/myApiPath/doHash",
      request({
        type: "myType1",
        name: "myName1",
        plainText: "example password",
      } satisfies DoHashRequest),
    );
    expect(mock).toHaveBeenNthCalledWith(
      2,
      "/myApiPath/doHash",
      request({
        type: "myType2",
        name: "myName2",
        plainText: "example password",
      } satisfies DoHashRequest),
    );
    expect(mock).toHaveBeenNthCalledWith(
      3,
      "/myApiPath/doHash",
      request({
        type: "myType3",
        name: "myName3",
        plainText: "example password",
      } satisfies DoHashRequest),
    );
    expect(mock).toHaveBeenCalledTimes(3);
    expect(
      await formatHtml(document.body.querySelector("tbody")!),
    ).toMatchSnapshot();
  });

  it("should render a table body with one hash measured", async () => {
    mockFetch().mockResolvedValueOnce(responseSuccess(threeHashes));
    await mountComponent();

    const mock = mockFetch();
    mock.mockResolvedValueOnce(
      responseSuccess({
        hash: "DO NOT ASSERT myHash2",
        elapsedNanos: 50020,
      } satisfies DoHashResponse),
    );
    (document.querySelectorAll(".measure").item(2) as HTMLElement).click();
    await flushPromises();
    expect(mock).toHaveBeenCalledExactlyOnceWith(
      "/myApiPath/doHash",
      request({
        type: "myType2",
        name: "myName2",
        plainText: "example password",
      } satisfies DoHashRequest),
    );
    expect(
      await formatHtml(document.body.querySelector("tbody")!),
    ).toMatchSnapshot();
  });

  it("should render a table body with one hash expanded", async () => {
    mockFetch().mockResolvedValueOnce(responseSuccess(threeHashes));
    await mountComponent();

    (document.querySelectorAll(".hash").item(1) as HTMLElement).click();
    await flushPromises();
    expect(
      await formatHtml(document.body.querySelector("tbody")!),
    ).toMatchSnapshot();

    const mock = mockFetch();
    mock.mockResolvedValueOnce(
      responseSuccess({
        hash: "myHash2",
        elapsedNanos: 50020,
      } satisfies DoHashResponse),
    );
    (
      document.querySelectorAll("td.expansion button").item(0) as HTMLElement
    ).click();
    await flushPromises();
    expect(mock).toHaveBeenCalledExactlyOnceWith(
      "/myApiPath/doHash",
      request({
        type: "myType2",
        name: "myName2",
        plainText: "",
      } satisfies DoHashRequest),
    );
    expect(
      await formatHtml(document.body.querySelector("tbody")!),
    ).toMatchSnapshot();
  });

  it("should render an error message", async () => {
    const mock = mockFetch();
    mock.mockResolvedValueOnce(responseFailure("myError"));
    await mountComponent();
    expect(mock).toHaveBeenCalledExactlyOnceWith("/myApiPath/hashes");
    expect(await formatHtml(document.body)).toMatchSnapshot();
  });
});

async function mountComponent() {
  mount(Hashes, { target: document.body }); // `document` exists because of jsdom
  await flushPromises();
}

const threeHashes: HashesResponse[] = [
  numberedHash(1),
  numberedHash(2),
  numberedHash(3),
];

function numberedHash(n: number): HashesResponse {
  return {
    type: "myType" + n,
    name: "myName" + n,
    plainTextLimit: 150 + n,
    plainTextValidator: undefined,
    algorithmID: "myAlgorithmID" + n,
    algorithmDescription: "myAlgorithmDescription" + n,
  };
}

beforeAll(() => {
  // https://github.com/jsdom/jsdom/issues/3429
  // @ts-expect-error
  Element.prototype.animate = () => ({});
});

afterAll(() => {
  // @ts-expect-error
  Element.prototype.animate = undefined;
});
