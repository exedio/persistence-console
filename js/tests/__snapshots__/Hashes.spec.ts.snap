// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Hashes > should render a non-empty table 1`] = `
"<table class="grey svelte-109884">
  <caption>
    Hashes
    <button class="reload svelte-dizlwz">🗘</button
    ><!---->
  </caption>
  <thead>
    <tr class="relative svelte-109884">
      <th rowspan="2">Type</th>
      <th rowspan="2">Name</th>
      <th colspan="2">Plain Text</th>
      <th colspan="2">Algorithm</th>
      <th rowspan="2" class="time svelte-109884">
        Time<small>/ns</small>
        <!----><button class="measure svelte-109884">🕐</button
        ><!---->
      </th>
    </tr>
    <tr>
      <th>Limit</th>
      <th>Validator</th>
      <th>ID</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
    <!---->
    <tr class="relative svelte-109884">
      <td>
        <button class="hash svelte-109884">↓<!----></button> myType
      </td>
      <td>myName</td>
      <td class="number">150</td>
      <td>myValidator</td>
      <td>myAlgorithmID</td>
      <td>myAlgorithmDescription</td>
      <td class="number">
        <!---->
        <button class="measure svelte-109884">🕐</button>
      </td>
    </tr>
    <!---->
    <tr class="relative svelte-109884">
      <td>
        <button class="hash svelte-109884">↓<!----></button> myTypeWithoutLimit
      </td>
      <td>myNameWithoutLimit</td>
      <td class="number">155</td>
      <td class="notAvailable"></td>
      <td>myAlgorithmIDWithoutLimit</td>
      <td>myAlgorithmDescriptionWithoutLimit</td>
      <td class="number">
        <!---->
        <button class="measure svelte-109884">🕐</button>
      </td>
    </tr>
    <!----><!---->
  </tbody>
</table>
<!---->
"
`;

exports[`Hashes > should render a non-empty table 2`] = `
"<table class="grey svelte-109884">
  <caption>
    Hashes
    <button class="reload svelte-dizlwz">🗘</button
    ><!---->
  </caption>
  <thead>
    <tr class="relative svelte-109884">
      <th rowspan="2">Type</th>
      <th rowspan="2">Name</th>
      <th colspan="2">Plain Text</th>
      <th colspan="2">Algorithm</th>
      <th rowspan="2" class="time svelte-109884">
        Time<small>/ns</small>
        <!----><button class="measure svelte-109884">🕐</button
        ><!---->
      </th>
    </tr>
    <tr>
      <th>Limit</th>
      <th>Validator</th>
      <th>ID</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
    <!---->
    <tr class="relative svelte-109884">
      <td>
        <button class="hash svelte-109884">↓<!----></button> myType
      </td>
      <td>myName</td>
      <td class="number">155</td>
      <td>myValidator</td>
      <td>myAlgorithmID</td>
      <td>myAlgorithmDescription</td>
      <td class="number">
        <!---->
        <button class="measure svelte-109884">🕐</button>
      </td>
    </tr>
    <!---->
    <tr class="relative svelte-109884">
      <td>
        <button class="hash svelte-109884">↓<!----></button> myTypeReloaded
      </td>
      <td>myNameReloaded</td>
      <td class="number">155</td>
      <td>myValidatorReloaded</td>
      <td>myAlgorithmIDReloaded</td>
      <td>myAlgorithmDescriptionReloaded</td>
      <td class="number">
        <!---->
        <button class="measure svelte-109884">🕐</button>
      </td>
    </tr>
    <!----><!---->
  </tbody>
</table>
<!---->
"
`;

exports[`Hashes > should render a table body with all hashes measured 1`] = `
"<!---->
<tr class="relative svelte-109884">
  <td>
    <button class="hash svelte-109884">↓<!----></button> myType1
  </td>
  <td>myName1</td>
  <td class="number">151</td>
  <td class="notAvailable"></td>
  <td>myAlgorithmID1</td>
  <td>myAlgorithmDescription1</td>
  <td class="number">
    50,010<!---->
    <button class="measure svelte-109884">🕐</button>
  </td>
</tr>
<!---->
<tr class="relative svelte-109884">
  <td>
    <button class="hash svelte-109884">↓<!----></button> myType2
  </td>
  <td>myName2</td>
  <td class="number">152</td>
  <td class="notAvailable"></td>
  <td>myAlgorithmID2</td>
  <td>myAlgorithmDescription2</td>
  <td class="number">
    50,020<!---->
    <button class="measure svelte-109884">🕐</button>
  </td>
</tr>
<!---->
<tr class="relative svelte-109884">
  <td>
    <button class="hash svelte-109884">↓<!----></button> myType3
  </td>
  <td>myName3</td>
  <td class="number">153</td>
  <td class="notAvailable"></td>
  <td>myAlgorithmID3</td>
  <td>myAlgorithmDescription3</td>
  <td class="number">
    50,030<!---->
    <button class="measure svelte-109884">🕐</button>
  </td>
</tr>
<!----><!---->
"
`;

exports[`Hashes > should render a table body with one hash expanded 1`] = `
"<!---->
<tr class="relative svelte-109884">
  <td>
    <button class="hash svelte-109884">↓<!----></button> myType1
  </td>
  <td>myName1</td>
  <td class="number">151</td>
  <td class="notAvailable"></td>
  <td>myAlgorithmID1</td>
  <td>myAlgorithmDescription1</td>
  <td class="number">
    <!---->
    <button class="measure svelte-109884">🕐</button>
  </td>
</tr>
<!---->
<tr class="relative svelte-109884">
  <td>
    <button class="hash svelte-109884">→<!----></button> myType2
  </td>
  <td>myName2</td>
  <td class="number">152</td>
  <td class="notAvailable"></td>
  <td>myAlgorithmID2</td>
  <td>myAlgorithmDescription2</td>
  <td class="number">
    <!---->
    <button class="measure svelte-109884">🕐</button>
  </td>
</tr>
<tr>
  <td colspan="7" class="expansion svelte-109884">
    <input placeholder="Plain Text" class="svelte-109884" />
    <button>Hash</button> <br />
    <small></small>
  </td>
</tr>
<!---->
<tr class="relative svelte-109884">
  <td>
    <button class="hash svelte-109884">↓<!----></button> myType3
  </td>
  <td>myName3</td>
  <td class="number">153</td>
  <td class="notAvailable"></td>
  <td>myAlgorithmID3</td>
  <td>myAlgorithmDescription3</td>
  <td class="number">
    <!---->
    <button class="measure svelte-109884">🕐</button>
  </td>
</tr>
<!----><!---->
"
`;

exports[`Hashes > should render a table body with one hash expanded 2`] = `
"<!---->
<tr class="relative svelte-109884">
  <td>
    <button class="hash svelte-109884">↓<!----></button> myType1
  </td>
  <td>myName1</td>
  <td class="number">151</td>
  <td class="notAvailable"></td>
  <td>myAlgorithmID1</td>
  <td>myAlgorithmDescription1</td>
  <td class="number">
    <!---->
    <button class="measure svelte-109884">🕐</button>
  </td>
</tr>
<!---->
<tr class="relative svelte-109884">
  <td>
    <button class="hash svelte-109884">→<!----></button> myType2
  </td>
  <td>myName2</td>
  <td class="number">152</td>
  <td class="notAvailable"></td>
  <td>myAlgorithmID2</td>
  <td>myAlgorithmDescription2</td>
  <td class="number">
    <!---->
    <button class="measure svelte-109884">🕐</button>
  </td>
</tr>
<tr>
  <td colspan="7" class="expansion svelte-109884">
    <input placeholder="Plain Text" class="svelte-109884" />
    <button>Hash</button> <br />
    <small>myHash2</small>
  </td>
</tr>
<!---->
<tr class="relative svelte-109884">
  <td>
    <button class="hash svelte-109884">↓<!----></button> myType3
  </td>
  <td>myName3</td>
  <td class="number">153</td>
  <td class="notAvailable"></td>
  <td>myAlgorithmID3</td>
  <td>myAlgorithmDescription3</td>
  <td class="number">
    <!---->
    <button class="measure svelte-109884">🕐</button>
  </td>
</tr>
<!----><!---->
"
`;

exports[`Hashes > should render a table body with one hash measured 1`] = `
"<!---->
<tr class="relative svelte-109884">
  <td>
    <button class="hash svelte-109884">↓<!----></button> myType1
  </td>
  <td>myName1</td>
  <td class="number">151</td>
  <td class="notAvailable"></td>
  <td>myAlgorithmID1</td>
  <td>myAlgorithmDescription1</td>
  <td class="number">
    <!---->
    <button class="measure svelte-109884">🕐</button>
  </td>
</tr>
<!---->
<tr class="relative svelte-109884">
  <td>
    <button class="hash svelte-109884">↓<!----></button> myType2
  </td>
  <td>myName2</td>
  <td class="number">152</td>
  <td class="notAvailable"></td>
  <td>myAlgorithmID2</td>
  <td>myAlgorithmDescription2</td>
  <td class="number">
    50,020<!---->
    <button class="measure svelte-109884">🕐</button>
  </td>
</tr>
<!---->
<tr class="relative svelte-109884">
  <td>
    <button class="hash svelte-109884">↓<!----></button> myType3
  </td>
  <td>myName3</td>
  <td class="number">153</td>
  <td class="notAvailable"></td>
  <td>myAlgorithmID3</td>
  <td>myAlgorithmDescription3</td>
  <td class="number">
    <!---->
    <button class="measure svelte-109884">🕐</button>
  </td>
</tr>
<!----><!---->
"
`;

exports[`Hashes > should render an empty table 1`] = `
"<table class="grey svelte-109884">
  <caption>
    Hashes
    <button class="reload svelte-dizlwz">🗘</button
    ><!---->
  </caption>
  <thead>
    <tr class="relative svelte-109884">
      <th rowspan="2">Type</th>
      <th rowspan="2">Name</th>
      <th colspan="2">Plain Text</th>
      <th colspan="2">Algorithm</th>
      <th rowspan="2" class="time svelte-109884">
        Time<small>/ns</small>
        <!----><!---->
      </th>
    </tr>
    <tr>
      <th>Limit</th>
      <th>Validator</th>
      <th>ID</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
    <!---->
    <tr>
      <td colspan="7" class="empty">There are no hashes in the model.</td>
    </tr>
    <!---->
  </tbody>
</table>
<!---->
"
`;

exports[`Hashes > should render an error message 1`] = `
"<table class="grey svelte-109884">
  <caption>
    Hashes
    <button class="reload svelte-dizlwz">🗘</button
    ><!---->
  </caption>
  <thead>
    <tr class="relative svelte-109884">
      <th rowspan="2">Type</th>
      <th rowspan="2">Name</th>
      <th colspan="2">Plain Text</th>
      <th colspan="2">Algorithm</th>
      <th rowspan="2" class="time svelte-109884">
        Time<small>/ns</small> &nbsp;<!---->
      </th>
    </tr>
    <tr>
      <th>Limit</th>
      <th>Validator</th>
      <th>ID</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td colspan="7" class="error">hashes: response code 400 myError</td>
    </tr>
    <!---->
  </tbody>
</table>
<!---->
"
`;
